lib/taro/rails/route_finder.rb

```diff
@@ -17,13 +17,11 @@ def cache
 
     def build_cache
       # Build a Hash like
-      # { 'users#show' } => [#<NormalizedRoute>, #<NormalizedRoute>] }
-      rails_routes.each_with_object({}) do |rails_route, hash|
-        route = Taro::Rails::NormalizedRoute.new(rails_route)
-        next if route.ignored?
-
-        (hash[route.endpoint] ||= []) << route
-      end
+      # { 'users#show' => [#<NormalizedRoute>, #<NormalizedRoute>] }
+      rails_routes
+        .map { |r| Taro::Rails::NormalizedRoute.new(r) }
+        .reject(&:ignored?)
+        .group_by { |r| r.endpoint }
```

Comment:
:+1:

i thought this could be slower because it goes over the routes three times instead of once, but it is actually faster, and quite fast in any case, even with 8000 routes:

```ruby
Benchmark.ips do |x|
  x.report('old') do
    rails_routes.each_with_object({}) do |rails_route, hash|
      route = Taro::Rails::NormalizedRoute.new(rails_route)
      next if route.ignored?
  
      (hash[route.endpoint] ||= []) << route
    end
  end
  x.report('new') do
    rails_routes
      .map { |r| Taro::Rails::NormalizedRoute.new(r) }
      .reject(&:ignored?)
      .group_by { |r| r.endpoint }
  end
end
```

```
ruby 3.3.4 (2024-07-09 revision be1089c8ec) [arm64-darwin23]
Warming up --------------------------------------
                 old     3.000 i/100ms
                 new     3.000 i/100ms
Calculating -------------------------------------
                 old     35.041 (± 2.9%) i/s   (28.54 ms/i) -    177.000 in   5.054543s
                 new     35.994 (± 2.8%) i/s   (27.78 ms/i) -    180.000 in   5.005303s
```

Reply:
Thanks for the benchmark. I've learned to appreciate [Enumerable](https://rubyapi.org/o/enumerable) :)

I'd also be interested to see the number of allocations. It could be that `group_by` is the biggest difference.

You could even try a third implementation:
```ruby
rails_routes.filter_map do |r|
  Taro::Rails::NormalizedRoute.new(r)
  r.ignored? ? nil : r
end.group_by(&:endpoint)
```
Coding style that's horrific and there are probably cleaner versions, but it may help avoiding to allocate an array twice.